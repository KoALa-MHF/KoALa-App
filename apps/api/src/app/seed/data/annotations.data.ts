import { Annotation } from '../../annotations/entities/annotation.entity';

export const AnnotationsData: Partial<Annotation>[] = [
  {
    id: 1,
    markerId: 1,
    userSessionId: 1,
    start: 200,
  },
];
