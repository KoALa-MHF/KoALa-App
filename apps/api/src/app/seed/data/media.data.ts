import { Media } from '../../media/entities/media.entity';

export const MediaData = [
  {
    id: 1,
    name: 'test.mp3',
    mimeType: 'audio/mpeg',
  },
] as [Media];
