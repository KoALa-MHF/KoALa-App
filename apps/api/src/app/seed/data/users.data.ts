export const UsersData = {
  sessionOwner1: {
    id: 1,
    displayName: 'Session Owner 1',
    email: 'session-owner-1@koala-app.de',
    samlId: '123456',
  },
  sessionOwner2: {
    id: 2,
    displayName: 'Session Owner 2',
    email: 'session-owner-2@koala-app.de',
    samlId: '654321',
  },
  sessionParticipant1: {
    id: 3,
    displayName: 'Session Participant 1',
    email: 'session-participant-1@koala-app.de',
  },
  sessionParticipant2: {
    id: 4,
    displayName: 'Session Participant 2',
    email: 'session-participant-2@koala-app.de',
  },
};
