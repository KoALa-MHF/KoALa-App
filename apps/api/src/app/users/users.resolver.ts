import { Resolver, Mutation, Args, Int, Query } from '@nestjs/graphql';
import { UsersService } from './users.service';
import { User } from './entities/user.entity';
import { UpdateUserInput } from './dto/update-user.input';
import { UseGuards } from '@nestjs/common';
import { AuthGuard } from '../core/guards/auth.guard';
import { CurrentUser } from '../core/decorators/user.decorator';

@Resolver(() => User)
export class UsersResolver {
  constructor(private readonly usersService: UsersService) {}

  @UseGuards(AuthGuard)
  @Query(() => User, { name: 'me' })
  me(@CurrentUser() user: User) {
    return this.usersService.findOne(user.id);
  }

  @UseGuards(AuthGuard)
  @Mutation(() => User)
  updateUser(
    @CurrentUser() user: User,
    @Args('updateUserInput')
    updateUserInput: UpdateUserInput
  ) {
    return this.usersService.update(id, updateUserInput);
  }
}
