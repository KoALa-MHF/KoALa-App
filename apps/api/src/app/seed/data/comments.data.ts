import { UsersData } from './users.data';

export const CommentsData = [
  {
    id: 1,
    owner: UsersData.sessionParticipant1,
    text: 'This is a comment from the Participant',
    annotationId: 1,
  },
  {
    id: 2,
    owner: UsersData.sessionOwner1,
    text: 'This is a comment from the Owner',
    annotationId: 1,
  },
];
