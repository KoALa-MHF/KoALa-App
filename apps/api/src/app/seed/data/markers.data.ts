import { MarkerType } from '../../markers/entities/marker.entity';
import { UsersData } from './users.data';

export const MarkersData = [
  {
    id: 1,
    type: MarkerType.EVENT,
    name: 'Marker 1',
    abbreviation: 'M1',
    description: 'Marker 1 Description',
    color: '#000000',
    icon: 'icon1',
    owner: UsersData.sessionOwner1,
    contentColor: '#FFFFFF',
  },
  {
    id: 2,
    type: MarkerType.RANGE,
    name: 'Marker 2',
    abbreviation: 'M2',
    description: 'Marker 2 Description',
    color: '#000000',
    icon: 'icon2',
    owner: UsersData.sessionOwner1,
    contentColor: '#FFFFFF',
  },
  {
    id: 3,
    type: MarkerType.SLIDER,
    name: 'Marker 3',
    abbreviation: 'M3',
    description: 'Marker 3 Description',
    color: '#000000',
    icon: 'icon3',
    rangeFrom: -5,
    rangeTo: 5,
    owner: UsersData.sessionOwner1,
    contentColor: '#FFFFFF',
  },
  {
    id: 4,
    type: MarkerType.EVENT,
    name: 'Marker 4',
    abbreviation: 'M4',
    description: 'Marker 4 Description',
    color: '#000000',
    icon: 'icon4',
    owner: UsersData.sessionOwner2,
    contentColor: '#FFFFFF',
  },
];
