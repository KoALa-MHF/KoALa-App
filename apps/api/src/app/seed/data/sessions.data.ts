import { SessionStatus } from '../../sessions/entities/session.entity';
import { UsersData } from './users.data';

export const SessionsData = [
  {
    id: 1,
    name: 'Session 1',
    owner: UsersData.sessionOwner1,
    status: SessionStatus.IN_PREPARATION,
  },
  {
    id: 2,
    name: 'Session 2',
    owner: UsersData.sessionOwner2,
    status: SessionStatus.CLOSED,
  },
  {
    id: 3,
    name: 'Session 3',
    owner: UsersData.sessionOwner1,
    status: SessionStatus.IN_PREPARATION,
    start: Date.now(),
    end: Date.now() + 1000000,
  },
  {
    id: 4,
    name: 'Session 4',
    owner: UsersData.sessionOwner1,
    status: SessionStatus.OPEN,
    start: Date.now() - 2000,
    end: Date.now() - 1000,
  },
  {
    id: 5,
    name: 'Session 5',
    owner: UsersData.sessionOwner1,
    status: SessionStatus.ARCHIVED,
  },
];
