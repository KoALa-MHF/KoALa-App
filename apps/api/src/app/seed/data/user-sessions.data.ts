import { UsersData } from './users.data';
import { SessionsData } from './sessions.data';

export const UserSessionsData = [
  {
    owner: UsersData.SESSION_PARTICIPANT_1,
    session: SessionsData[0],
  },
  {
    owner: UsersData.SESSION_PARTICIPANT_2,
    session: SessionsData[0],
  },
  {
    owner: UsersData.SESSION_OWNER_1,
    session: SessionsData[1],
  },
  {
    owner: UsersData.SESSION_PARTICIPANT_2,
    session: SessionsData[2],
  },
];
