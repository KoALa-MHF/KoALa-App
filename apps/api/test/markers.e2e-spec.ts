import request from 'supertest-graphql';
import gql from 'graphql-tag';
import { INestApplication } from '@nestjs/common';
import { setupTestApplication } from './mocks/test.util';
import { UsersData } from '../src/app/seed/data/users.data';
import { MarkerType } from '../src/app/markers/entities/marker.entity';

const QUERY_MARKERS = gql`
  query Markers {
    markers {
      id
      name
      description
      type
      color
      icon
    }
  }
`;

const CREATE_MARKER = gql`
  mutation CreateMarker($createMarkerInput: CreateMarkerInput!) {
    createMarker(createMarkerInput: $createMarkerInput) {
      id
      name
      description
      type
      color
      icon
    }
  }
`;

const UPDATE_MARKER = gql`
  mutation UpdateMarker($id: Int!, $updateMarkerInput: UpdateMarkerInput!) {
    updateMarker(id: $id, updateMarkerInput: $updateMarkerInput) {
      id
      name
      description
      type
      color
      icon
    }
  }
`;

const REMOVE_MARKER = gql`
  mutation RemoveMarker($id: Int!) {
    removeMarker(id: $id) {
      id
      name
      description
      type
      color
      icon
    }
  }
`;

const CREATE_MARKER_VARIABLES = {
  createMarkerInput: {
    name: 'Event Marker',
    description: 'Event Marker Description',
    type: 'EVENT',
    color: '#000000',
    icon: 'icon1',
  },
};

const UPDATE_MARKER_VARIABLES = {
  id: 1,
  updateMarkerInput: {
    name: 'Event Marker Updated',
    description: 'Event Marker Description Updated',
    type: 'EVENT',
    color: '#FFFFFF',
    icon: 'icon2',
  },
};

const REMOVE_MARKER_VARIABLES = {
  id: 2,
};

describe('Markers (e2e)', () => {
  let app: INestApplication;

  beforeEach(async () => {
    app = await setupTestApplication();
  });

  afterEach(async () => {
    await app.close();
  });

  describe('Query Markers', () => {
    it('Not authenticated user should get "Unauthorized" error', async () => {
      const { errors } = await request(app.getHttpServer()).query(QUERY_MARKERS);

      expect(errors).toHaveLength(1);
      expect(errors[0].message).toBe('Unauthorized');
    });

    it('Session owner should get list of all markers, even of markers of other owners', async () => {
      const { data } = await request(app.getHttpServer())
        .auth(`${UsersData.sessionOwner1.id}`, { type: 'bearer' })
        .query(QUERY_MARKERS)
        .expectNoErrors();

      expect(data).toMatchSnapshot();
    });

    it('Participating user should get list of all markers', async () => {
      const { data } = await request(app.getHttpServer())
        .auth(`${UsersData.sessionParticipant1.id}`, { type: 'bearer' })
        .query(QUERY_MARKERS)
        .expectNoErrors();

      expect(data).toMatchSnapshot();
    });
  });

  describe('Create Markers', () => {
    it('Not authenticated user should get "Unauthorized" error', async () => {
      const { errors } = await request(app.getHttpServer()).mutate(CREATE_MARKER).variables(CREATE_MARKER_VARIABLES);

      expect(errors).toHaveLength(1);
      expect(errors[0].message).toBe('Unauthorized');
    });

    it('SAML Authenticated user can create a new marker', async () => {
      const { data } = await request(app.getHttpServer())
        .auth(`${UsersData.sessionOwner1.id}`, { type: 'bearer' })
        .mutate(CREATE_MARKER)
        .variables(CREATE_MARKER_VARIABLES)
        .expectNoErrors();

      expect(data).toMatchSnapshot();
    });

    it('None SAML Authenticated user cannot create a new marker', async () => {
      const { errors } = await request(app.getHttpServer())
        .auth(`${UsersData.sessionParticipant1.id}`, { type: 'bearer' })
        .mutate(CREATE_MARKER)
        .variables(CREATE_MARKER_VARIABLES);

      expect(errors).toHaveLength(1);
      expect(errors[0].message).toBe('Forbidden resource');
    });
  });

  describe('Update Marker', () => {
    it('Not authenticated user should get "Unauthorized" error', async () => {
      const { errors } = await request(app.getHttpServer()).mutate(UPDATE_MARKER).variables(UPDATE_MARKER_VARIABLES);

      expect(errors).toHaveLength(1);
      expect(errors[0].message).toBe('Unauthorized');
    });

    it('Marker Owner can update a marker', async () => {
      const { data } = await request(app.getHttpServer())
        .auth(`${UsersData.sessionOwner1.id}`, { type: 'bearer' })
        .mutate(UPDATE_MARKER)
        .variables(UPDATE_MARKER_VARIABLES)
        .expectNoErrors();

      expect(data).toMatchSnapshot();
    });

    it('Marker Owner of another marker cannot update a marker and should get "Not Found" error', async () => {
      const { errors } = await request(app.getHttpServer())
        .auth(`${UsersData.sessionOwner2.id}`, { type: 'bearer' })
        .mutate(UPDATE_MARKER)
        .variables(UPDATE_MARKER_VARIABLES);

      expect(errors).toHaveLength(1);
      expect(errors[0].message).toBe('Not Found');
    });

    it('Participant (None SAML Authenticated user) cannot update a marker and should get "Forbidden resource" error', async () => {
      const { errors } = await request(app.getHttpServer())
        .auth(`${UsersData.sessionParticipant1.id}`, { type: 'bearer' })
        .mutate(UPDATE_MARKER)
        .variables(UPDATE_MARKER_VARIABLES);

      expect(errors).toHaveLength(1);
      expect(errors[0].message).toBe('Forbidden resource');
    });
  });

  describe('Remove Marker', () => {
    it('Not authenticated user should get "Unauthorized" error', async () => {
      const { errors } = await request(app.getHttpServer()).mutate(REMOVE_MARKER).variables(REMOVE_MARKER_VARIABLES);

      expect(errors).toHaveLength(1);
      expect(errors[0].message).toBe('Unauthorized');
    });

    it('Marker Owner can remove a marker', async () => {
      const { data: sessionsBeforeDelete } = await request(app.getHttpServer())
        .auth(`${UsersData.sessionOwner1.id}`, { type: 'bearer' })
        .query(QUERY_MARKERS)
        .expectNoErrors();
      expect(sessionsBeforeDelete).toMatchSnapshot();

      const { data } = await request(app.getHttpServer())
        .auth(`${UsersData.sessionOwner1.id}`, { type: 'bearer' })
        .mutate(REMOVE_MARKER)
        .variables(REMOVE_MARKER_VARIABLES)
        .expectNoErrors();
      expect(data).toMatchSnapshot();

      const { data: sessionsAfterDelete } = await request(app.getHttpServer())
        .auth(`${UsersData.sessionOwner1.id}`, { type: 'bearer' })
        .query(QUERY_MARKERS)
        .expectNoErrors();
      expect(sessionsAfterDelete).toMatchSnapshot();
    });

    it('Marker Owner of another marker cannot remove the marker and should get "Not Found" error', async () => {
      const { errors } = await request(app.getHttpServer())
        .auth(`${UsersData.sessionOwner2.id}`, { type: 'bearer' })
        .mutate(REMOVE_MARKER)
        .variables(REMOVE_MARKER_VARIABLES);

      expect(errors).toHaveLength(1);
      expect(errors[0].message).toBe('Not Found');
    });

    it('Participant (None SAML Authenticated user) cannot remove a session and should get "Forbidden resource" error', async () => {
      const { errors } = await request(app.getHttpServer())
        .auth(`${UsersData.sessionParticipant1.id}`, { type: 'bearer' })
        .mutate(REMOVE_MARKER)
        .variables(REMOVE_MARKER_VARIABLES);

      expect(errors).toHaveLength(1);
      expect(errors[0].message).toBe('Forbidden resource');
    });
  });
});
