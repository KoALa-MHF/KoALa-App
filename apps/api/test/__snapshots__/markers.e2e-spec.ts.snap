// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Markers (e2e) Create Markers SAML Authenticated user can create a new marker 1`] = `
{
  "createMarker": {
    "color": "#000000",
    "description": "Event Marker Description",
    "icon": "icon1",
    "id": 5,
    "name": "Event Marker",
    "type": "EVENT",
  },
}
`;

exports[`Markers (e2e) Query Markers Participating user should get list of all markers 1`] = `
{
  "markers": [
    {
      "color": "#000000",
      "description": "Marker 1 Description",
      "icon": "icon1",
      "id": 1,
      "name": "Marker 1",
      "type": "EVENT",
    },
    {
      "color": "#000000",
      "description": "Marker 2 Description",
      "icon": "icon2",
      "id": 2,
      "name": "Marker 2",
      "type": "RANGE",
    },
    {
      "color": "#000000",
      "description": "Marker 3 Description",
      "icon": "icon3",
      "id": 3,
      "name": "Marker 3",
      "type": "SLIDER",
    },
    {
      "color": "#000000",
      "description": "Marker 4 Description",
      "icon": "icon4",
      "id": 4,
      "name": "Marker 4",
      "type": "EVENT",
    },
  ],
}
`;

exports[`Markers (e2e) Query Markers Session owner should get list of all markers, even of markers of other owners 1`] = `
{
  "markers": [
    {
      "color": "#000000",
      "description": "Marker 1 Description",
      "icon": "icon1",
      "id": 1,
      "name": "Marker 1",
      "type": "EVENT",
    },
    {
      "color": "#000000",
      "description": "Marker 2 Description",
      "icon": "icon2",
      "id": 2,
      "name": "Marker 2",
      "type": "RANGE",
    },
    {
      "color": "#000000",
      "description": "Marker 3 Description",
      "icon": "icon3",
      "id": 3,
      "name": "Marker 3",
      "type": "SLIDER",
    },
    {
      "color": "#000000",
      "description": "Marker 4 Description",
      "icon": "icon4",
      "id": 4,
      "name": "Marker 4",
      "type": "EVENT",
    },
  ],
}
`;

exports[`Markers (e2e) Remove Marker Marker Owner can remove a marker 1`] = `
{
  "markers": [
    {
      "color": "#000000",
      "description": "Marker 1 Description",
      "icon": "icon1",
      "id": 1,
      "name": "Marker 1",
      "type": "EVENT",
    },
    {
      "color": "#000000",
      "description": "Marker 2 Description",
      "icon": "icon2",
      "id": 2,
      "name": "Marker 2",
      "type": "RANGE",
    },
    {
      "color": "#000000",
      "description": "Marker 3 Description",
      "icon": "icon3",
      "id": 3,
      "name": "Marker 3",
      "type": "SLIDER",
    },
    {
      "color": "#000000",
      "description": "Marker 4 Description",
      "icon": "icon4",
      "id": 4,
      "name": "Marker 4",
      "type": "EVENT",
    },
  ],
}
`;

exports[`Markers (e2e) Remove Marker Marker Owner can remove a marker 2`] = `
{
  "removeMarker": {
    "color": "#000000",
    "description": "Marker 2 Description",
    "icon": "icon2",
    "id": 2,
    "name": "Marker 2",
    "type": "RANGE",
  },
}
`;

exports[`Markers (e2e) Remove Marker Marker Owner can remove a marker 3`] = `
{
  "markers": [
    {
      "color": "#000000",
      "description": "Marker 1 Description",
      "icon": "icon1",
      "id": 1,
      "name": "Marker 1",
      "type": "EVENT",
    },
    {
      "color": "#000000",
      "description": "Marker 3 Description",
      "icon": "icon3",
      "id": 3,
      "name": "Marker 3",
      "type": "SLIDER",
    },
    {
      "color": "#000000",
      "description": "Marker 4 Description",
      "icon": "icon4",
      "id": 4,
      "name": "Marker 4",
      "type": "EVENT",
    },
  ],
}
`;

exports[`Markers (e2e) Update Marker Marker Owner can update a marker 1`] = `
{
  "updateMarker": {
    "color": "#FFFFFF",
    "description": "Event Marker Description Updated",
    "icon": "icon2",
    "id": 1,
    "name": "Event Marker Updated",
    "type": "EVENT",
  },
}
`;
